# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1722981412 -7200
# Parent  28ca9a2cebd3370407085baf5e21cca5576525df
Feature: Follow-up to bug 1908761 - partial backout to fix secondary sort in grouped-by views.

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -7525,10 +7525,16 @@ var sortController = {
     // differently than other variations.
     // So, first set the desired sortType and sortOrder, then set viewFlags in
     // batch mode, then apply it all (open a new view) with endViewUpdate().
     gViewWrapper.beginViewUpdate();
+    // Note: this.dbView still remembers the last secondary sort, before group
+    // sort was entered. If we do not specify a secondary sort here, dbView.open()
+    // will use the new primary sort and the old (!) secondary sort. Let's push
+    // the current primary sort as the new secondary sort.
+    const [curPrimarySort] = gViewWrapper._sort;
     gViewWrapper._sort = [
       [newSortType, Ci.nsMsgViewSortOrder.ascending, newSortColumnId],
+      curPrimarySort,
     ];
     gViewWrapper.showGroupedBySort = false;
     gViewWrapper.endViewUpdate();
     // Virtual folders don't persist viewFlags well in the back end,
diff --git a/mail/modules/DBViewWrapper.sys.mjs b/mail/modules/DBViewWrapper.sys.mjs
--- a/mail/modules/DBViewWrapper.sys.mjs
+++ b/mail/modules/DBViewWrapper.sys.mjs
@@ -34,10 +34,9 @@ function getSortStatusFromView(dbView) {
     secondarySort[0] == Ci.nsMsgViewSortType.byNone ||
     (secondarySort[0] != Ci.nsMsgViewSortType.byCustom &&
       secondarySort[0] == primarySort[0]) ||
     (secondarySort[0] == Ci.nsMsgViewSortType.byCustom &&
-      secondarySort[2] == primarySort[2]) ||
-    dbView.viewFlags & Ci.nsMsgViewFlagsType.kGroupBySort
+      secondarySort[2] == primarySort[2])
   ) {
     return [primarySort];
   }
   return [primarySort, secondarySort];
diff --git a/mailnews/base/src/nsMsgDBView.cpp b/mailnews/base/src/nsMsgDBView.cpp
--- a/mailnews/base/src/nsMsgDBView.cpp
+++ b/mailnews/base/src/nsMsgDBView.cpp
@@ -4170,16 +4170,11 @@ nsresult nsMsgDBView::RestoreSortInfo() 
     nsString sortColumnsString;
     folderInfo->GetProperty("sortColumns", sortColumnsString);
     DecodeColumnSort(sortColumnsString);
     if (m_sortColumns.Length() > 1) {
-      if (m_viewFlags & nsMsgViewFlagsType::kGroupBySort) {
-        // Discard any persisted secondary sort information.
-        m_sortColumns.RemoveElementAt(1);
-      } else {
-        m_secondarySort = m_sortColumns[1].mSortType;
-        m_secondarySortOrder = m_sortColumns[1].mSortOrder;
-        m_secondaryCustomColumn = m_sortColumns[1].mCustomColumnName;
-      }
+      m_secondarySort = m_sortColumns[1].mSortType;
+      m_secondarySortOrder = m_sortColumns[1].mSortOrder;
+      m_secondaryCustomColumn = m_sortColumns[1].mCustomColumnName;
     }
 
     // Restore curCustomColumn from db.
     folderInfo->GetProperty("customSortCol", m_curCustomColumn);
